src/policy/marksweepspace/native_ms/block.rs

```diff
@@ -0,0 +1,447 @@
+        if cfg!(feature = "malloc_native_mimalloc") {
+            unimplemented!()
+        }
+
+        // Check if we can treat it as the simple case: cell address === object reference.
+        // If the binding does not use allocation offset, and they use the same allocation alignment which the cell size is aligned to,
+        // then we have cell address === allocation address.
+        // Furthermore, if the binding does not have an offset between allocation and object reference, then allocation address === cell address.
+        if !VM::USE_ALLOCATION_OFFSET
+            && VM::MAX_ALIGNMENT == VM::MIN_ALIGNMENT
+            && crate::util::conversions::raw_is_aligned(
+                self.load_block_cell_size(),
+                VM::MAX_ALIGNMENT,
+            )
+            && VM::VMObjectModel::OBJECT_REF_OFFSET_LOWER_BOUND == 0
+            && VM::VMObjectModel::OBJECT_REF_OFFSET_UPPER_BOUND == 0
+        {
+            // In this case, we can use the simplest and the most efficicent sweep.
+            self.simple_sweep::<VM>()
+        } else {
+            // Otherwise we fallback to a generic but slow sweep. This roughly has ~10% mutator overhead for lazy sweeping.
+            self.naive_brute_force_sweep::<VM>()
+        }
+    }
+
+    /// This implementation uses object reference and cell address interchangably. This is not correct for most cases.
+    /// However, in certain cases, such as OpenJDK, this is correct, and efficient. See the sweep method for the invariants
+    /// that we need to use this method correctly.
+    fn simple_sweep<VM: VMBinding>(&self) {
+        let cell_size = self.load_block_cell_size();
+        let mut cell = self.start();
+        let mut last = unsafe { Address::zero() };
+        while cell + cell_size <= self.start() + Block::BYTES {
+            // We know the cell and the object reference is the same.
+            if !VM::VMObjectModel::LOCAL_MARK_BIT_SPEC
+                .is_set::<VM>(unsafe { cell.to_object_reference() }, Ordering::SeqCst)
+            {
+                // clear alloc bit if it is ever set.
+                #[cfg(feature = "global_alloc_bit")]
+                crate::util::alloc_bit::ALLOC_SIDE_METADATA_SPEC.store_atomic::<u8>(
+                    cell,
+                    0,
+                    Ordering::SeqCst,
+                );
+                unsafe {
+                    cell.store::<Address>(last);
+                }
+                last = cell;
+            }
+            cell += cell_size;
+        }
+
+        self.store_free_list(last);
+    }
+
+    /// This is a naive implementation that is inefficient but should be correct.
+    /// In this implementation, we simply go through each possible object
+    /// reference and see if it has the mark bit set. If we find mark bit, that means the cell is alive. If we didn't find
+    /// the mark bit in the entire cell, it means the cell is dead.
+    fn naive_brute_force_sweep<VM: VMBinding>(&self) {
```

Comment:
Or we can work out the in-cell offset given any address inside that object.
```rust
let any_address = ObjectModel::ref_to_cell(objref);
let in_cell_offset = (any_address - block_start) % cell_size;
```
Then we know the start address of an allocation (which is equal to the start of the cell in the freelist allocator) by
```rust
let start_of_cell = any_address - in_cell_offset;
assert_eq!(start_of_cell, ObjectModel::object_start_ref(objref));
```
In this way, we only need to do arithmetic (and avoid accessing object headers) to know the cell start.  Then we can reuse the per-word mark bit side-metadata.

Reply:
> I wonder if we can optimise this if `ObjectModel` provides a function that is guaranteed to return any address within an object given an object reference (currently `ObjectModel::ref_to_address`). If each cell has the same size, then given any address within an object, we can workout the cell index by dividing the in-block offset by the cell size. In this way, we don't need a per-object mark bit, but a per-cell mark bit local to each block.

What you said would work. It is conceptually simpler. But it has some drawbacks:
1. We already have an object mark bit (in header or on side), and we do not want to introduce other metadata for the cells.
2. If we do not use object mark bit, and use the cell bit instead, it might be slower than what this code is doing. Cell size for each block is stored in side metadata. To get the cell address, we need to load the cell size, and then do computation. If this happens for any query on the object liveness, it could be slow as well.
3. This 'naive' method is the worst case we could have (which has ~10-15% mutator overhead), but the simplest method (`simple_sweep()`) which works for OpenJDK, there is no overhead. Between those two cases, we still have chance to improve.

Reply:
> Or we can work out the in-cell offset given any address inside that object.
> 
> ```rust
> let any_address = ObjectModel::ref_to_cell(objref);
> let in_cell_offset = (any_address - block_start) % cell_size;
> ```
> 
> Then we know the start address of an allocation (which is equal to the start of the cell in the freelist allocator) by
> 
> ```rust
> let start_of_cell = any_address - in_cell_offset;
> assert_eq!(start_of_cell, ObjectModel::object_start_ref(objref));
> ```
> 
> In this way, we only need to do arithmetic (and avoid accessing object headers) to know the cell start. Then we can reuse the per-word mark bit side-metadata.

The address we return from `alloc()` is based on the alignment and offset for that allocation. The `in_cell_offset` is not a constant. We can only optimize if we know some invariants, like the conditional checks for using `simple_sweep()`.